File: sdk/typescript/src/api/keys.ts
import type { HttpClient } from "../http.js";
import type {
  KeyBundle,
  KeyHealth,
  PreKeysRequest,
  SignedPreKeyRequest,
} from "../types/index.js";

export class KeysApi {
  constructor(private readonly http: HttpClient) {}

  getBundle(agentId: string): Promise<KeyBundle> {
    return this.http.get<KeyBundle>(
      `/keys/${encodeURIComponent(agentId)}/bundle`,
    );
  }

  health(agentId: string): Promise<KeyHealth> {
    return this.http.getDirectoryAuth<KeyHealth>(
      `/keys/${encodeURIComponent(agentId)}/health`,
    );
  }

  uploadPreKeys(agentId: string, request: PreKeysRequest): Promise<void> {
    return this.http.putDirectoryAuth<void>(
      `/keys/${encodeURIComponent(agentId)}/prekeys`,
      request,
    );
  }

  rotateSignedPreKey(
    agentId: string,
    request: SignedPreKeyRequest,
  ): Promise<void> {
    return this.http.putDirectoryAuth<void>(
      `/keys/${encodeURIComponent(agentId)}/signed-prekey`,
      request,
    );
  }
}

File: sdk/rust/src/api/keys.rs
use std::sync::Arc;

use crate::error::Result;
use crate::http::{HttpClient, to_body};
use crate::types::{KeyBundle, PreKeysRequest, SignedPreKeyRequest};

pub struct KeysApi {
    http: Arc<HttpClient>,
}

impl KeysApi {
    pub fn new(http: Arc<HttpClient>) -> Self {
        Self { http }
    }

    pub async fn get_bundle(&self, agent_id: &str) -> Result<KeyBundle> {
        let path = format!("/keys/{}/bundle", urlencoding::encode(agent_id));
        self.http.get_auth(&path, None).await
    }

    pub async fn upload_pre_keys(
        &self,
        agent_id: &str,
        request: &PreKeysRequest,
    ) -> Result<()> {
        let path = format!("/keys/{}/prekeys", urlencoding::encode(agent_id));
        self.http.put(&path, Some(&to_body(request)?)).await
    }

    pub async fn rotate_signed_pre_key(
        &self,
        agent_id: &str,
        request: &SignedPreKeyRequest,
    ) -> Result<()> {
        let path = format!("/keys/{}/signed-prekey", urlencoding::encode(agent_id));
        self.http.put(&path, Some(&to_body(request)?)).await
    }
}
